### DISCUSSION TITLE
Implementing Babystep-Giantstep Matrix-Vector Multiplication with BFV scheme

### DISCUSSION THREAD

**Discussion by fedejinich** [tuneinsight/lattigo#360] [Q&A] [answered]
Hi, I'm trying to transcipher from PASTA to BFV, and one of the key elements to achieve this is to evaluate a product between a plain matrix and an encrypted packed vector. The [original PASTA paper](https://eprint.iacr.org/2021/731.pdf) suggests using the Babystep-Gigantstep algorithm and provides a [reference](https://github.com/IAIK/hybrid-HE-framework/blob/d8bf788ce119bae577dc5ed500b45ad3392f66c2/ciphers/pasta_3/seal/pasta_3_seal.cpp#L163) implementation. I'm trying to reproduce it in Go, using `lattigo` instead of [SEAL](https://github.com/microsoft/SEAL)

![Screenshot 2023-04-11 at 18 38 41](https://user-images.githubusercontent.com/11727247/231293961-e19632b5-68fa-497b-916e-efdd58b1532c.png)

- Does anybody know if this is achievable with lattigo? 
- Can I rotate rows by 'n' steps in the same [way as SEAL does](https://github.com/microsoft/SEAL/blob/206648d0e4634e5c61dcf9370676630268290b59/native/src/seal/evaluator.h#L1053)?

**Comment by Pro7ech** [comment, marked as answer]
Hi @fedejinich,

The algorithm is implemented for the BGV scheme: [`linear_transforms.go`](https://github.com/tuneinsight/lattigo/blob/master/bgv/linear_transforms.go). Reimplementing it for BFV should be straightforward. 

Another option is the experimental branch [`https://github.com/tuneinsight/lattigo/tree/dev_bgvfv`](https://github.com/tuneinsight/lattigo/tree/dev_bgvfv) which implements a unification of the BFV and BGV scheme. In this branch the BFV sheme is instantiated as a wrapper of a generalization of BGV, thus it has the API for the linear transform (see [`bfv.go`](https://github.com/tuneinsight/lattigo/blob/dev_bgvfv/bfv/bfv.go) and [this line](https://github.com/tuneinsight/lattigo/blob/fe82ce93dee57bb42afbcc8f2084811963bac86a/bfv/bfv_test.go#L669) for an example of how to use it).

**Comment by fedejinich** [comment]
Hi @Pro7ech thanks for the quick (and nice) response, It really helps me. 
I'm trying to implement a hybrid holomorphic scheme (PASTA + BFV) and right now this is one of my main blockers. 
I guess I'll experiment with both options and see the results.

Again, thank you :)

**Comment by fedejinich** [comment]
@Pro7ech Do you think the experimental branch will be included in a new version? If not, I can try to collaborate with another experimental branch with the BFV implementation

**Reply by Pro7ech** [reply]
@fedejinich Great :) I'm glad this helped. Yes this branch is PR#309, which is third in the line of the PRs that are going to be merged once they are reviewed.

**Reply by fedejinich** [reply]
okay, thank you very much!
